internal/mux/buffer_test.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 The Oto Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mux_test
+
+import (
+	"io"
+	"sync"
+	"testing"
+
+	"github.com/hajimehoshi/oto/internal/mux"
+)
+
+const concurrency = 1000
+
+func TestBufferConcurrentWrites(t *testing.T) {
+	b := &mux.ConcurrentBuffer{}
+
+	var done, start sync.WaitGroup
+	done.Add(concurrency)
+	start.Add(1)
+	for i := 0; i < concurrency; i++ {
+		go func() {
+			start.Wait()
+			b.Write([]byte{1})
+			done.Done()
+		}()
+	}
+	start.Done()
+	done.Wait()
+
+	if l := b.Len(); l != concurrency {
+		t.Errorf("b.Len: got: %v, want: %v", l, concurrency)
+	}
+
+	out := make([]byte, concurrency)
+	if _, err := io.ReadFull(b, out); err != nil {
+		t.Fatal(err)
+	}
+
+	for _, b := range out {
+		if b != 1 {
+			t.Errorf("expected to find all 1s in the buffer, but there was a '%v'", b)
+			break
+		}
+	}
+}
+
+func TestConcurrentReadWrites(t *testing.T) {
```

Comment:
I think you are trying to test the weak synchronization, right? To test the weak synchronization, we'd need a test that tests multiple players. However, I think this is very hard to test it.

If not, I'm not sure this test would be worth: I thought this was a test for mutex.

What do you think?

Reply:
This is a test that would fail on simply a `bytes.Buffer`. We need `Read` and `Write` from different goroutines to block one another, otherwise we end up losing data. This test was not necessary with `io.Pipe` because it handled that for us. But now that we have our own data structure we should test it works as we need.

So, I think this should stay.

Reply:
Ah i see, thanks!

> This test was not necessary with io.Pipe because it handled that for us. But now that we have our own data structure we should test it works as we need.

That's true, so this test tests the mutex, right?

I'm fine with having this test anyway, thank you for elaborating!

Reply:
If you had the question, someone else down the line would have the same question. Always happy to elaborate in a PR to give context to those who come after me. :)